hal/tests/TESTS/pin_names/pinvalidate.py

```diff
@@ -14,29 +14,62 @@
 limitations under the License.
 """
 
-import os
-import re
 import argparse
-import pathlib
 import json
+import os
+import pathlib
+import re
+import sys
 from tabulate import tabulate
+from enum import Enum
+
+
+class ReturnCode(Enum):
+    """Return codes."""
+
+    SUCCESS = 0
+    ERROR = 1
+    INVALID_OPTIONS = 2
+
+
+class TestCaseError(Exception):
+    """An exception for test case failure."""
+
+
+class ArgumentParserWithDefaultHelp(argparse.ArgumentParser):
+    """Subclass that always shows the help message on invalid arguments."""
+
+    def error(self, message):
+        """Error handler."""
+        sys.stderr.write("error: {}\n".format(message))
+        self.print_help()
+        raise SystemExit(ReturnCode.INVALID_OPTIONS.value)
+
+
+def find_target(target_name=""):
+    """Find a target."""
+    mbed_os_root = os.path.abspath(
+        os.path.join(
+            pathlib.Path(__file__).parent.absolute(), "..", "..", "..", ".."
+        )
+    )
 
-def find_target(target_name = ''):
-    mbed_os_root = os.path.abspath(os.path.join(pathlib.Path(__file__).parent.absolute(), '..', '..', '..', '..'))
     targets = dict()
-    
-    try:
-        targets_json_file = open(os.path.join(mbed_os_root, 'targets', 'targets.json'))
-        target_data = json.load(targets_json_file)
-    except:
-        exit("Failed to open targets.json")
 
-    for root, dirs, files in os.walk(mbed_os_root):
+    targets_json_file = open(
+        os.path.join(mbed_os_root, "targets", "targets.json")
+    )
+
+    target_data = json.load(targets_json_file)
+
+    for root, _, files in os.walk(mbed_os_root):
         for f in files:
             if f == "PinNames.h":
                 path = os.path.join(root, f)
-                name_components = re.findall("TARGET_([a-zA-Z0-9_]*)[\/\\\]", path)
-                name = '_'.join(name_components)
+                name_components = re.findall(
+                    "TARGET_([a-zA-Z0-9_]*)[\/\\\]", path
+                )
+                name = "_".join(name_components)
                 if not name:
                     continue
                 if target_name:
```

Comment:
If I understand this block relative to the previous one correctly, you are filtering your list of names based on the regex. I think it might be cleaner to create a "private" function (i.e. one starting with a single underscore) for your filter that operated on the Path objects. Then you could do something like this:

```python
def _filter_target_names(p: Path) -> bool:
    return is_this_a_valid_target_or_not_regex_wizardry

### now in this function

for f in filter(_filter_target_names, mbed_os_root.rglob("PinNames.h")):
   ### magic!
```

Here, `filter` takes the single item output of your generator and only does something if the first argument (called with the item) returns true.

Reply:
That would be neat, but the output of the regex/join is not only used to filter names but also as a fallback value in case a target cannot be matched 1:1 with one from targets.json.